onnx/defs/nn/defs.cc

```diff
@@ -2487,4 +2487,195 @@ ONNX_OPERATOR_SET_SCHEMA(
         )ONNX"
         ));
 
+void col2imShapeInference(InferenceContext& ctx) {
+  propagateElemTypeFromInputToOutput(ctx, 0, 0);
+
+  // validating 'pads' attribute
+  std::vector<int64_t> pads;
+  if (getRepeatedAttribute(ctx, "pads", pads)) {
+    if (pads.size() != 2 * 2) {
+      fail_shape_inference("Attribute pads has incorrect size");
+    }
+  }
+
+  std::vector<int64_t> dilations;
+  if (getRepeatedAttribute(ctx, "dilations", dilations)) {
+    if (dilations.size() != 2) {
+      return;
+    }
+  }
+
+  std::vector<int64_t> strides;
+  if (getRepeatedAttribute(ctx, "strides", strides)) {
+    if (strides.size() != 2) {
+      return;
+    }
+  }
+
+  // validating 'block_shape' input
+  std::vector<int64_t> block_shape;
+  const TensorProto* temp_block_shape = ctx.getInputData(2);
+  if (temp_block_shape) {
+    block_shape = ParseData<int64_t>(temp_block_shape);
+    if (block_shape.size() != 2) {
+      fail_shape_inference("block_shape tensor has wrong rank. It must have rank 2.");
+    }
+  } else {
+    if (ctx.getInputType(2)->tensor_type().shape().dim_size() != 1) {
+      fail_shape_inference("block_shape tensor has wrong rank. It must have rank 2.");
+    } else if (ctx.getInputType(2)->tensor_type().shape().dim(0).dim_value() != 2) {
+      fail_shape_inference("block_shape tensor has wrong rank. It must have rank 2.");
+    }
+  }
+
+
+  // we need at least two inputs to have a shape for this inference.
+  if (!hasNInputShapes(ctx, 2)) {
+    return;
+  }
+
+  // validating 'input' input
+  auto input_shape = ctx.getInputType(0)->tensor_type().shape();
+  if ((input_shape.dim_size() < 2) || (input_shape.dim_size() > 3)) {
+    fail_shape_inference("input tensor has wrong rank. It must have rank 2 or 3.");
+  }
+
+  // validating 'image_shape' input
+  std::vector<int64_t> image_shape;
+  const TensorProto* temp_image_shape = ctx.getInputData(1);
+  if (temp_image_shape) {
+    image_shape = ParseData<int64_t>(temp_image_shape);
+    if (image_shape.size() != 2) {
+      fail_shape_inference("image_shape tensor has wrong rank. It must have rank 2.");
+    }
+  } else {
+    if (ctx.getInputType(1)->tensor_type().shape().dim_size() != 1) {
+      fail_shape_inference("image_shape tensor has wrong rank. It must have rank 2.");
+    } else if (ctx.getInputType(1)->tensor_type().shape().dim(0).dim_value() != 2) {
+      fail_shape_inference("image_shape tensor has wrong rank. It must have rank 2.");
+    }
+  }
+
+  // inferencing final shape
+  auto final_image_shape =
+      ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
+
+  auto size_of_output = static_cast<int>(image_shape.size());
+  // a new dim is always added to the output
+  final_image_shape->add_dim()->set_dim_value(1);
```

Comment:
Don't understand this. Why is this 1? And at the beginning?

Reply:
Is there some pseudo-code description of what the op does? That would help understand and verify this shape-inference.

Reply:
I think this is supposed to be batch-size, if input is unbatched? Not sure it is a good idea. If the input is unbatched (2D), then it seems better for output to be unbatched (3D).

Reply:
> Is there some pseudo-code description of what the op does? That would help understand and verify this shape-inference.

Im2col transforms several patches in one big matrix
![image](https://user-images.githubusercontent.com/5469809/159054523-1330c8df-eb03-42fb-bf98-98ff5edca406.png)

while col2im transforms the big matrix big in the original shape
![image](https://user-images.githubusercontent.com/5469809/159054662-a0d375bc-0050-4ad9-976f-74c05520271e.png)

Reply:
great catch. The first output dimension is the input's batch dimension, which defaults to 1 for unbatched input. Fixed.